File: archivist/src/main/java/com/zorroa/archivist/web/api/PluginController.java
package com.zorroa.archivist.web.api;

import com.google.common.collect.ImmutableMap;
import com.zorroa.archivist.domain.Plugin;
import com.zorroa.archivist.domain.Processor;
import com.zorroa.archivist.domain.ProcessorFilter;
import com.zorroa.archivist.service.PluginService;
import com.zorroa.common.config.ApplicationProperties;
import com.zorroa.sdk.exception.PluginException;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;

/**
 * Created by chambers on 6/29/16.
 */

@RestController
public class PluginController {

    @Autowired
    ApplicationProperties properties;

    @Autowired
    PluginService pluginService;

    @ResponseBody
    @RequestMapping(value="/api/v1/plugins", method = RequestMethod.POST)
    public Object handlePluginUpload(@RequestParam("file") MultipartFile file) {
        if (!file.isEmpty()) {
            if (!file.getOriginalFilename().endsWith("-plugin.zip")) {
                throw new RuntimeException("The plugin package name must end with -plugin.zip");
            }
            Plugin p = pluginService.installPlugin(file);
            return ImmutableMap.of(
                    "name",p.getName(),
                    "description", p.getDescription(),
                    "version", p.getVersion(),
                    "publisher", p.getPublisher(),
                    "language", p.getLanguage());
        }
        else {
            throw new PluginException("Failed to handle plugin zip file");
        }
    }

    @RequestMapping(value="/api/v1/processors", method=RequestMethod.GET)
    public List<Processor> processors(@RequestBody(required = false) ProcessorFilter filter) {
        if (filter == null) {
            return pluginService.getAllProcessors();
        }
        else {
            return pluginService.getAllProcessors(filter);
        }
    }
}

File: archivist/src/main/kotlin/com/zorroa/archivist/web/api/PluginController.kt
package com.zorroa.archivist.web.api

import com.google.common.collect.ImmutableMap
import com.zorroa.archivist.domain.Processor
import com.zorroa.archivist.domain.ProcessorFilter
import com.zorroa.archivist.service.PluginService
import com.zorroa.archivist.util.KJson.UUID_REGEXP
import com.zorroa.sdk.client.exception.ArchivistWriteException
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.web.bind.annotation.*
import org.springframework.web.multipart.MultipartFile
import java.util.*

/**
 * Created by chambers on 6/29/16.
 */

@PreAuthorize("hasAuthority(T(com.zorroa.archivist.sdk.security.Groups).DEV) || hasAuthority(T(com.zorroa.archivist.sdk.security.Groups).ADMIN)")
@RestController
class PluginController @Autowired constructor(
        private val pluginService: PluginService
){

    @ResponseBody
    @RequestMapping(value = ["/api/v1/plugins"], method = [RequestMethod.POST])
    fun handlePluginUpload(@RequestParam("file") file: MultipartFile): Any {
        if (!file.isEmpty) {
            if (!file.originalFilename.endsWith("-plugin.zip")) {
                throw RuntimeException("The plugin package name must end with -plugin.zip")
            }
            val p = pluginService.installPlugin(file)
            return ImmutableMap.of(
                    "name", p.name,
                    "description", p.description,
                    "version", p.version,
                    "publisher", p.publisher,
                    "language", p.language)
        } else {
            throw ArchivistWriteException("Failed to handle plugin zip file")
        }
    }

    @RequestMapping(value = ["/api/v1/processors"], method = [RequestMethod.GET, RequestMethod.POST])
    fun getProcessors(@RequestBody(required = false) filter: ProcessorFilter?): List<Processor> {
        return if (filter == null) {
            pluginService.getAllProcessors()
        } else {
            pluginService.getAllProcessors(filter)
        }
    }

    @GetMapping(value = "/api/v1/processors/{id:.+}")
    fun getProcessor(@PathVariable id: String): Processor {
        return if (UUID_REGEXP.matches(id)) {
            pluginService.getProcessor(UUID.fromString(id))
        }
        else {
            pluginService.getProcessor(id)
        }
    }
}
